internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+}
+
+func MeasureEndpoint(ctx context.Context, creq *CtrlRequest, URL *url.URL, endpoint string, wg *sync.WaitGroup, out chan *MeasureEndpointResult) {
+	defer wg.Done()
+	endpointResult := &MeasureEndpointResult{}
+	measureFactory[URL.Scheme](ctx, creq, endpoint, wg, endpointResult)
+	out <- endpointResult
+}
+
+var measureFactory = map[string]func(ctx context.Context, creq *CtrlRequest, endpoint string, wg *sync.WaitGroup, result *MeasureEndpointResult){
+	"http":  measureHTTP,
+	"https": measureHTTP,
+	"h3":    measureH3,
+	"h3-29": measureH3,
+}
+
+func measureHTTP(
+	ctx context.Context,
+	creq *CtrlRequest,
+	endpoint string,
+	wg *sync.WaitGroup,
+	result *MeasureEndpointResult,
+) {
+	URL, _ := url.Parse(creq.HTTPRequest)
+	httpMeasurement := CtrlHTTPMeasurement{Endpoint: endpoint, Protocol: URL.Scheme}
+	var conn net.Conn
+	conn, httpMeasurement.TCPConnect = TCPDo(ctx, &TCPConfig{
+		Endpoint: endpoint,
+	})
+	if conn == nil {
+		return
+	}
+	defer conn.Close()
+	var transport http.RoundTripper
+	switch URL.Scheme {
+	case "http":
+		transport = nwebconnectivity.GetSingleTransport(conn, nil)
+	case "https":
+		var tlsconn *tls.Conn
+		cfg := &tls.Config{ServerName: URL.Hostname()}
+		tlsconn, httpMeasurement.TLSHandshake = TLSDo(ctx, &TLSConfig{
+			Conn:     conn,
+			Endpoint: endpoint,
+			Cfg:      cfg,
+		})
+		if tlsconn == nil {
+			return
+		}
+		transport = nwebconnectivity.GetSingleTransport(tlsconn, cfg)
+	}
+	// perform the HTTP request: this provides us with the HTTP request result and info about HTTP redirection
+	httpMeasurement.HTTPRequest, result.httpRedirect = HTTPDo(ctx, &HTTPConfig{
+		Jar:       creq.HTTPCookieJar,
+		Headers:   creq.HTTPRequestHeaders,
+		Transport: transport,
+		URL:       URL,
+	})
+	// find out of the host also supports h3 support, which is announced in the Alt-Svc Header
+	h3Support := discoverH3Server(httpMeasurement.HTTPRequest, URL)
+	if h3Support != "" {
+		quicURL, _ := url.Parse(URL.String())
+		quicURL.Scheme = h3Support
+		result.h3Location = quicURL.String()
+	}
+	result.CtrlEndpoint = &httpMeasurement
+}
+
+func measureH3(
+	ctx context.Context,
+	creq *CtrlRequest,
+	endpoint string,
+	wg *sync.WaitGroup,
+	result *MeasureEndpointResult,
+) {
+	URL, _ := url.Parse(creq.HTTPRequest)
+	h3Measurement := CtrlH3Measurement{Endpoint: endpoint, Protocol: URL.Scheme}
+	var sess quic.EarlySession
+	tlscfg := &tls.Config{ServerName: URL.Hostname(), NextProtos: []string{URL.Scheme}}
+	qcfg := &quic.Config{}
+	sess, h3Measurement.QUICHandshake = QUICDo(ctx, &QUICConfig{
+		Endpoint:  endpoint,
+		QConfig:   qcfg,
+		TLSConfig: tlscfg,
+	})
+	if sess == nil {
+		return
+	}
+	transport := nwebconnectivity.GetSingleH3Transport(sess, tlscfg, qcfg)
+	h3Measurement.HTTPRequest, result.httpRedirect = HTTPDo(ctx, &HTTPConfig{
+		Jar:       creq.HTTPCookieJar,
+		Headers:   creq.HTTPRequestHeaders,
+		Transport: transport,
+		URL:       URL,
+	})
+	sess.CloseWithError(0, "")
```

Comment:
Is this the idiomatic way to close an HTTP3 session? Is this the right error code? (Can't we ask the transport to do this thing for us instead?)